File: Blind75/Array Product Excluding Current/Array_Product_Excluding_Current.js
// Brute Force 
function arrayProductExcludingCurrent(numbers){
    const length = numbers.length
    let products = Array(length).fill(1)
    
    for (let i = 0; i < length; i++){
        for (let j = 0; j < length; j++){
            if(i !== j){ 
                products[i] *= numbers[j] 
            }
        }
    }

    return products
}

// Optimal Approach
function arrayProductExcludingCurrent(numbers){
    const length = numbers.length
    let answer  = Array(length).fill(1)

    let prefix = 1
    for (let i = 0; i < length; i++){
        answer[i] = prefix
        prefix = prefix * numbers[i]
    }

    let suffix = 1
    for (let i = length - 1; i >= 0; i--){
        answer[i] = answer[i] *  suffix
        suffix = suffix * numbers[i]
    }
    
    return answer
}

const testCases = [[1,2,3], [2,0,3], [0,0,-1,1]]
for(const testCase of testCases){
    console.log(arrayProductExcludingCurrent(testCase))
}

File: Blind75/Array Product Excluding Current/Array_Product_Excluding_Current.py
def arrayProductExcludingCurrent(numbers : list[int]) -> list[int]:
    length = len(numbers)
    answer = []

    for i in range(length):
        product = 1
        for j in range(length):
            if j != i:
                product *= numbers[j]
        answer.append(product)

    return answer


test_cases = [[1,2,3], [2,0,3], [0,0,-1,1]]
for test_case in test_cases:
    print(arrayProductExcludingCurrent(test_case))